fast_llm/layers/decoder/stochastic_mixer.py

```diff
@@ -0,0 +1,189 @@
+            config,
+            distributed_config,
+            hidden_dim=hidden_dim,
+            lr_scale=lr_scale,
+            peft=peft,
+            return_bias=return_bias,
+        )
+
+        # Initialize all mixers
+        self.mixers = torch.nn.ModuleDict(
+            {
+                name: mixer_config.get_layer(
+                    distributed_config,
+                    hidden_dim,
+                    lr_scale,
+                    peft=peft,
+                    return_bias=return_bias,
+                )
+                for name, mixer_config in self._config.mixers.items()
+            }
+        )
+
+        # Precompute sampling probabilities as a tensor (ordered by mixers.keys())
+        if self._config.sampling_strategy == SamplingStrategy.uniform:
+            self._sampling_probs = torch.ones(len(self.mixers)) / len(self.mixers)
+        elif self._config.sampling_strategy == SamplingStrategy.weighted:
+            if self._config.sampling_weights is None:
+                raise ValueError("sampling_weights must be provided when using weighted sampling strategy")
+            self._sampling_probs = torch.tensor(
+                [self._config.sampling_weights[name] for name in self.mixers.keys()], dtype=torch.float32
+            )
+        else:
+            raise NotImplementedError(f"Sampling strategy {self._config.sampling_strategy} not implemented")
+
+        logger.info(
+            f"Initialized StochasticMixer with {len(self.mixers)} mixers: "
+            f"{', '.join(f'{name}={type(mixer).__name__}' for name, mixer in self.mixers.items())} "
+            f"(main={self._config.main_mixer_name})"
+        )
+
+        # Mark all mixer parameters with allow_no_grad since only one mixer
+        # is active per forward pass during training. Even though all mixers
+        # will eventually be trained, on any single forward pass, the non-selected
+        # mixers won't receive gradients.
+        for mixer in self.mixers.values():
+            for param in mixer.parameters(recurse=True):
+                if hasattr(param, "allow_no_grad"):
+                    param.allow_no_grad = True
+
+    def setup(self, distributed: Distributed) -> None:
+        """Setup all mixers with the distributed context."""
+        super().setup(distributed)
+        for mixer in self.mixers.values():
+            mixer.setup(distributed)
+
+    def _sample_mixer_name(self) -> str:
+        """
+        Sample a mixer name according to the configured strategy.
+        In debug mode, verifies all ranks in the TP/PP group sample the same index.
+
+        Returns:
+            Name of the mixer to use for this forward pass.
+        """
+        if not self.training:
+            # Use main mixer for inference
+            return self._config.main_mixer_name
+
+        # Sample index in training mode
+        generator = self._distributed.tp_generator if self._sequence_parallel else self._distributed.pp_generator
```

Comment:
I don't think that's right, `tp_generator` will result in different tensor ranks selecting different mixers.